### PR TITLE
Squashed AIRBIN generation and runtime

### PR DESCRIPTION
**Pull request by jgmelber**
This PR adds initial runtime support for AIRBINs from [mlir-aie #243](https://github.com/Xilinx/mlir-aie/pull/243). AIRBIN generation and use is demonstrated in test/301_airbin_gen_add_one.

### CODE REVIEW

File: runtime_lib/airhost/CMakeLists.txt [hunk 1]
```diff
@@ -67,5 +69,5 @@ foreach(file ${INSTALLS})
 endforeach()
 
 # Install files
-set(INSTALLS memory.cpp queue.cpp host.cpp utility.cpp pcie-ernic.cpp pcie-ernic-dev-mem-allocator.cpp network.cpp)
+set(INSTALLS airbin.cpp memory.cpp queue.cpp host.cpp utility.cpp pcie-ernic.cpp pcie-ernic-dev-mem-allocator.cpp network.cpp)
```
Comment (fifield) [inline runtime_lib/airhost/CMakeLists.txt:72]: I think we can remove this line for good in a subsequent PR.

File: runtime_lib/airhost/airbin.cpp [hunk 2]
```diff
@@ -0,0 +1,292 @@
+
```
Comment (fifield) [inline runtime_lib/airhost/airbin.cpp:1]: Needs license

File: runtime_lib/airhost/airbin.cpp [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+struct Air64_Chdr {
+  uint16_t ch_tile;
+  uint8_t ch_name;
+  uint32_t ch_type;
+  uint64_t ch_addr;
+  uint64_t ch_offset;
+  uint64_t ch_size;
+};
+
+std::string to_air_cfg_name(uint32_t in) {
```
Comment (fifield) [inline runtime_lib/airhost/airbin.cpp:54]: make this static?

File: runtime_lib/airhost/include/airbin.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#ifndef AIRBIN_H
```
Comment (fifield) [inline runtime_lib/airhost/include/airbin.h:1]: needs license

File: runtime_lib/airhost/include/airbin.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+struct airbin_size {
+  uint8_t start_col = 0;
+  uint8_t num_cols = 1;
+  uint8_t start_row = 1;
+  uint8_t num_rows = 2;
+};
+
+airbin_size readairbinsize(std::ifstream &infile, uint8_t column_offset);
+void readairbin(std::ifstream &infile);
```
Comment (fifield) [inline runtime_lib/airhost/include/airbin.h:19]: How does this relate to `air_load_airbin`? It should be a C API if it gets `#included` in `air_host.h` which is supposed to be a C header*. The alternative is to `#include` it in `air.hpp` which is much less desirable if these functions are required to use airbin.

\*  and `<string>` has already crept in

File: test/278_portable_airbin_config/Makefile [hunk 6]
```diff
@@ -0,0 +1,15 @@
+.PHONY: all
```
Comment (fifield) [inline test/278_portable_airbin_config/Makefile:1]: needs license

File: test/278_portable_airbin_config/test.cpp [hunk 7]
```diff
@@ -0,0 +1,877 @@
+// (c) Copyright 2021 Xilinx Inc. All Rights Reserved.
```
Comment (fifield) [inline test/278_portable_airbin_config/test.cpp:1]: needs updated license

File: test/279_dual_herd/test.cpp [hunk 8]
```diff
@@ -0,0 +1,845 @@
+// (c) Copyright 2021 Xilinx Inc. All Rights Reserved.
```
Comment (fifield) [inline test/279_dual_herd/test.cpp:1]: needs updated license

File: test/300_relocatable_airbin_config/Makefile [hunk 9]
```diff
@@ -1,9 +1,8 @@
-# Copyright (C) 2022, Xilinx Inc.
-# Copyright (C) 2022, Advanced Micro Devices, Inc.
-# SPDX-License-Identifier: MIT
```
Comment (fifield) [inline test/300_relocatable_airbin_config/Makefile:3]: undelete license

File: test/300_relocatable_airbin_config/test.cpp [hunk 10]
```diff
@@ -0,0 +1,262 @@
+// (c) Copyright 2021 Xilinx Inc. All Rights Reserved.
```
Comment (fifield) [inline test/300_relocatable_airbin_config/test.cpp:1]: needs updated license

File: test/301_airbin_gen_add_one/Makefile [hunk 11]
```diff
@@ -0,0 +1,41 @@
+ACDC_AIE = $(dir $(shell which aie-opt))/..
```
Comment (fifield) [inline test/301_airbin_gen_add_one/Makefile:1]: needs license

File: test/301_airbin_gen_add_one/test.cpp [hunk 12]
```diff
@@ -0,0 +1,157 @@
+// (c) Copyright 2021 Xilinx Inc. All Rights Reserved.
```
Comment (fifield) [inline test/301_airbin_gen_add_one/test.cpp:1]: needs updated license

**Review by fifield — CHANGES_REQUESTED** [review]
I'm curious why we're checking in `.airbin` files instead of generating them?

File: runtime_lib/airhost/airbin.cpp [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+  config_header.ch_tile = (location_and_name >> 8u) & 0xFFFFu;
+  config_header.ch_tile += static_cast<uint16_t>(column_offset) << 5u;
+
+  assert(config_header.ch_tile != 0);
+  assert(config_header.ch_size % 4 == 0);
+
+  return config_header;
+}
+
+airbin_size readairbinsize(std::ifstream &infile, uint8_t column_offset) {
```
Comment (stephenneuendorffer) [inline runtime_lib/airhost/airbin.cpp:136]: camelcase or underscores?

File: runtime_lib/airhost/airbin.cpp [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+        return lhs.second < rhs.second;
+      });
+
+  result.start_row = minmax_rows.first->second;
+  result.num_rows = (minmax_rows.second->second - result.start_row) + 1u;
+
+  return result;
+}
+
+void readairbin(std::ifstream &infile) {
```
Comment (stephenneuendorffer) [inline runtime_lib/airhost/airbin.cpp:170]: underscores?

File: runtime_lib/airhost/airbin.cpp (hunk 13 above)
Comment (fifield) [reply runtime_lib/airhost/airbin.cpp:136]: static?
+1 for underscores

File: runtime_lib/airhost/airbin.cpp [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+      std::cout << std::hex << std::setw(6) << std::setfill('0')
+                << config_header.ch_offset << ' ';
+      std::cout << std::hex << std::setw(6) << std::setfill('0')
+                << config_header.ch_size << std::endl;
+    }
+    next_chcfg_idx++;
+  }
+}
+
+uint64_t airbin2mem(std::ifstream &infile, volatile uint32_t *tds_va,
```
Comment (fifield) [inline runtime_lib/airhost/airbin.cpp:240]: static?

File: runtime_lib/airhost/airbin.cpp (hunk 14 above)
Comment (fifield) [reply runtime_lib/airhost/airbin.cpp:170]: static?

**Review by fifield — APPROVED** [review]

**Comment by jgmelber** [comment]
Work continued by #259

**Pull request closed without merging**